### PR TITLE
systemd: add withLogTrace option

### PR DESCRIPTION
**Pull request by JulienMalka**
## Description of changes

This option is very useful for debugging systemd

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ElvishJerricco** [comment]
Can you explain what this does? The description in systemd's `meson_options.txt` isn't very helpful:

> enable low level debug logging

**Comment by JulienMalka** [comment]
> Can you explain what this does? The description in systemd's `meson_options.txt` isn't very helpful:
> 
> > enable low level debug logging

Yes, basically `systemd` runs with a high debug level in the console. Useful to debug situations where the `systemd` messages are unclear, it helped me debug credentials problems.

**Comment by ElvishJerricco** [comment]
Is that different than `systemd.log_level=debug` on the cmdline?

File: pkgs/os-specific/linux/systemd/default.nix [hunk 1]
```diff
@@ -142,6 +142,7 @@
 , docbook_xsl
 , docbook_xml_dtd_42
 , docbook_xml_dtd_45
+, withLogTrace ? false
```
Comment (AndersonTorres) [inline pkgs/os-specific/linux/systemd/default.nix:145]: why not make it true?

Comment (flokli) [reply pkgs/os-specific/linux/systemd/default.nix:145]: This causes `LOG_TRACE` to be defined, which then definess `log_trace*` functions actually doing logging.

I assume some of these are quite heavy perf-wise, so they're not compiled in by default.

**Review by AndersonTorres — APPROVED** [review]

**Review by ElvishJerricco — CHANGES_REQUESTED** [review]
> Is that different than systemd.log_level=debug on the cmdline?

This PR should not be approved without an answer to this question. We don't need more systemd build options when cmdline will do.

**Comment by JulienMalka** [comment]
Yes, this is doing more than systemd.log_level=debug on the cmdline.
> This causes LOG_TRACE to be defined, which then definess log_trace* functions actually doing logging.

**Review by ElvishJerricco — APPROVED** [review]
@JulienMalka Ah, so `log_trace*` is like *even higher* logging than debug. But it ordinarily is disabled not just at runtime, but code isn't even generated for this log level? In that case I think this makes perfect sense. Thanks :)

**Comment by JulienMalka** [comment]
> @JulienMalka Ah, so `log_trace*` is like _even higher_ logging than debug. But it ordinarily is disabled not just at runtime, but code isn't even generated for this log level? In that case I think this makes perfect sense. Thanks :)

That is exactly that, yes

**Comment by AndersonTorres** [comment]
> This PR should not be approved without an answer to this question. We don't need more systemd build options when cmdline will do.

It was implicitly answered here:  https://github.com/NixOS/nixpkgs/pull/262952/files#r1403135897

> This causes LOG_TRACE to be defined, which then definess log_trace* functions actually doing logging.

**Pull request merged**